### PR TITLE
Add security module: Verify messages on fetch

### PR DESCRIPTION
**Pull request by MHHukiewitz**
EDIT: to be merged after #136 

- Added `security.py` that offers a simple catch-all solution for verifying messages & posts.
- Added `verify_signatures` param to most methods of `AlephClient` to automatically verify signatures, if requested (False by default)
- Added some missing parameters to the `get_posts_iterator` and `get_messages_iterator` functions that were present on their non-iterator cousins.
- Renamed `sol.py` to `solana.py` but still import all of solana.py in sol.py to keep backwards compatibility.
- Renamed `polkadot` extra dependency to `substrate` to align with the module name in `aleph.sdk`.

### CODE REVIEW

File: src/aleph/sdk/client/abstract.py [hunk 1]
```diff
@@ -75,25 +76,35 @@ async def get_posts(
         :param post_filter: Filter to apply to the posts (Default: None)
         :param ignore_invalid_messages: Ignore invalid messages (Default: True)
         :param invalid_messages_log_level: Log level to use for invalid messages (Default: logging.NOTSET)
+        :param verify_signatures: Whether to verify the signatures of the messages (Default: False)
```
Comment (hoh) [inline src/aleph/sdk/client/abstract.py:79]: Would it be more readable in the imperative form ?
```suggestion
        :param verify_signatures: Verify the signatures of the messages (Default: False)
```

File: src/aleph/sdk/client/abstract.py [hunk 2]
```diff
@@ -75,25 +76,35 @@ async def get_posts(
...
+        verify_signatures: bool = False,
     ) -> AsyncIterable[PostMessage]:
         """
         Fetch all filtered posts, returning an async iterator and fetching them page by page. Might return duplicates
         but will always return all posts.
 
         :param post_filter: Filter to apply to the posts (Default: None)
+        :param ignore_invalid_messages: Ignore invalid messages (Default: True)
+        :param invalid_messages_log_level: Log level to use for invalid messages (Default: logging.NOTSET)
+        :param verify_signatures: Whether to verify the signatures of the messages (Default: False)
```
Comment (hoh) [inline src/aleph/sdk/client/abstract.py:97]: Would it be more readable in the imperative form ?
```suggestion
        :param verify_signatures: Verify the signatures of the messages (Default: False)
```

File: src/aleph/sdk/client/abstract.py [hunk 3]
```diff
@@ -168,25 +180,35 @@ async def get_messages(
         :param message_filter: Filter to apply to the messages
         :param ignore_invalid_messages: Ignore invalid messages (Default: True)
         :param invalid_messages_log_level: Log level to use for invalid messages (Default: logging.NOTSET)
+        :param verify_signatures: Whether to verify the signatures of the messages (Default: False)
```
Comment (hoh) [inline src/aleph/sdk/client/abstract.py:183]: Would it be more readable in the imperative form ?
```suggestion
        :param verify_signatures: Verify the signatures of the messages (Default: False)
```

File: src/aleph/sdk/security.py [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+from typing import Any, Union
+
+from aleph_message.models import AlephMessage
+
+from aleph.sdk.chains.common import get_verification_buffer
+from aleph.sdk.query.responses import Post
+
+validator_chains_map = {
+    # TODO: Add AVAX
+    "ETH": "ethereum",
```
Comment (hoh) [inline src/aleph/sdk/security.py:11]: We already have chain abbreviations defined here: https://github.com/aleph-im/aleph-message/blob/main/aleph_message/models/base.py#L4-L16

```python
from aleph_message import Chain
...
    Chain.ETH: "ethereum",
```

File: src/aleph/sdk/security.py [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+
+def try_import_verify_signature(chain: str) -> Any:
+    """Try to import a chain signature validator."""
+    try:
+        return import_module(f"aleph.sdk.chains.{chain}").verify_signature
+    except (ImportError, AttributeError):
+        return None
+
+
+validators = {
```
Comment (hoh) [inline src/aleph/sdk/security.py:28]: Can you explain what this is for in a comment ?

File: src/aleph/sdk/security.py [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+    key: try_import_verify_signature(value)
+    for key, value in validator_chains_map.items()
+}
+
+
+def verify_message_signature(message: Union[AlephMessage, Post]) -> None:
+    """Verify the signature of a message."""
```
Comment (hoh) [inline src/aleph/sdk/security.py:35]: Can you be more detailed ? Repeating the name of the function is not very useful...

```suggestion
    """Verify the signature of a message, raise an error if invalid or unsupported.
    A BadSignatureError is raised when the signature is incorrect.
    A ValueError is raised when the chain is not supported or required dependencies are  missing.
    """
```

**Comment by MHHukiewitz** [comment]
Why the the hell is the ubuntu image trying to install the .NET SDK and why is it breaking our CI/CD? :/

File: src/aleph/sdk/security.py [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+        raise ValueError(f"Chain {message.chain} is not supported.")
+    validator = validators[message.chain]
+    if validator is None:
+        raise ValueError(
+            f"Chain {message.chain} is not installed. Install it with `aleph-sdk-python[{message.chain}]`."
+        )
+    signature = message.signature
+    public_key = message.sender
+    message = get_verification_buffer(message.dict())
+    validator(signature, public_key, message)
```
Comment (nesitor) [inline src/aleph/sdk/security.py:54]: Can you also add a test case for that method?

Comment (MHHukiewitz) [reply src/aleph/sdk/security.py:54]: I don't see the value in it. It is a simple glue method that takes already well-tested functionality (`get_verification_buffer` and each chain's according `verify_signature` function, all tested) and just does a check on whether the requested chain is available.

**Comment by Psycojoker** [comment]
To be merged after #136

**Comment by Psycojoker** [comment]
Reading this PR it seems pretty harmless to merge since everything is turned off by default. Which brings the question of: when do we want to activate the verification by default?

On the other hand I'm not really happy with the test suit, the existing one hasn't been adapted to turn the verification on :/

**Comment by Psycojoker** [comment]
Current status conclusion after a discussion with Hugo:

* globally this PR looks ok on the principle
* but the tests aren't good enough
* in the tests, we need to do a better test on the `aleph.sdk.security.verify_message_signature` function
* in the tests, we need to launch `aleph.sdk.client.http.*` method call with `verify_signatures = True` which is currently not done

**Comment by MHHukiewitz** [comment]
I'm sorry, but am I expected to finish this? @Psycojoker Maybe you could look into it and incorporate the aforementioned feedback.

**Comment by Psycojoker** [comment]
> I'm sorry, but am I expected to finish this? @Psycojoker Maybe you could look into it and incorporate the aforementioned feedback.

@MHHukiewitz not that I know, Hugo asked me to at least to a quick pass to bring back this PR and clean it to see if it's mergeable because he told me that it seemed important. We just concluded that we wanted more tests before merging it.

**Pull request closed without merging**